### ISSUE TITLE
Grid context menu issues with falsey record ids

### ISSUE THREAD

**Issue by haynesjm42** [exhi/hoist-react#488]
Encountered this in a client app where the data I put in the store for the grid does not have an `id` property, so `LocalStore` sets it to the index in the raw data, which results in the first record having an `id` of `0`.

What happens in this case is that when the grid component tries to determine the record that was clicked on for the context menu it is implicitly converting the selected record id to a boolean instead of explicitly checking for `null` or `undefined` (lodash `isNil()`), so the `record` parameter passed to the `StoreContextMenuItem` callbacks is `null`.

It was very easy to workaround the issue in my case (just setting an `id` prop on the data to `i + 1` before loading it into the store) but we might want to do a quick audit of code that deals with record ids to make sure we are handling falsey ids properly

**Comment by amcclain** [comment]
Thanks for filing this ticket.  Always want to be on the lookout for falsey bugs and aggressively hunt them down.  For clarity - is this line w/the bug? Anything else involved?

https://github.com/exhi/hoist-react/blob/0dc739d2ddeeff0ba5aa8235afd9ee07e57bb085/desktop/cmp/grid/Grid.js#L159

**Sorry to piggy back on this ticket**, but I also want to check in re. what we consider a valid record ID and discuss how / where we could document and validate that.  An ID of 0 certainly seems valid (if a bit unusual), so I do think we should allow it, but empty string seems like a bad idea, as does a bool, other object, etc.

Also, looking at the code in `LocalStore` to auto-create IDs, it seems potentially problematic to [use the index of the record](https://github.com/exhi/hoist-react/blob/0dc739d2ddeeff0ba5aa8235afd9ee07e57bb085/data/LocalStore.js#L114). Doesn't seem impossible that you could have some records with IDs and others without, and the index of a rec w/o an ID overlaps with an already-assigned ID...  ExtJS maintains a global seed value it always increments and prefixes generated IDs to reliably produce unique values. We could take the same approach.

**Comment by haynesjm42** [comment]
Yep that's the line.

I believe we discussed the record id/index thing months back because of some use-cases I have in a client app where I am often putting "new" records into a store before they have been persisted (and thus do not yet have an id), and I ended up overriding some of the behavior in LocalStore with my own base Store class similar to what you mentioned ExtJs does (though it's not very robust currently). I haven't really thought about it much since I had a solution in place for the client app.

I think it makes a lot of sense to get away from using the `id` property of records internally since it is almost always used as the identity column name, `__xhId` or something that is obviously for internal use and guaranteed to be unique across stores would give us more flexibility for the data we can put into stores.
